pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartialUpsertHandler.java

```diff
@@ -32,76 +31,79 @@
-              newRecord.removeNullValueField(column);
-            } else if (!_comparisonColumns.contains(column)) {
-              newRecord.putValue(column, merger.merge(prevValue, newRecord.getValue(column)));
-            }
-          }
-        } else {
-          // Overwrite mergers.
-          // (1) If the merge strategy is Overwrite merger and newValue is not null, skip and use the new value
-          // (2) Otherwise, if previous is not null, init columnReader and use the previous value.
-          if (newRecord.isNullValue(column)) {
-            Object prevValue = prevRecord.getValue(column);
-            if (prevValue != null) {
-              newRecord.putValue(column, prevValue);
-              newRecord.removeNullValueField(column);
-            }
-          }
+  public void merge(LazyRow prevRecord, GenericRow newRecord, Map<String, Object> reuseMergerResult) {
+    reuseMergerResult.clear();
+
+    // merger current row with previously indexed row
+    _partialUpsertMerger.merge(prevRecord, newRecord, reuseMergerResult);
+
+    if (_partialUpsertMerger instanceof PartialUpsertColumnarMerger) {
+      // iterate over all columns in prevRecord and update newRecord with merged values
+      for (String column : prevRecord.getColumnNames()) {
+        // no merger to apply on primary key columns
+        if (_primaryKeyColumns.contains(column) || _comparisonColumns.contains(column)) {
+          continue;
+        }
+
+        // use merged column value from result map
+        if (reuseMergerResult.containsKey(column)) {
+          Object mergedValue = reuseMergerResult.get(column);
+          setMergedValue(newRecord, column, mergedValue);
+        }
+      }
+    } else {
+      // iterate over only merger results and update newRecord with merged values
+      for (Map.Entry<String, Object> entry : reuseMergerResult.entrySet()) {
+        // skip if primary key column
+        String column = entry.getKey();
+        if (_primaryKeyColumns.contains(column) || _comparisonColumns.contains(column)) {
+          continue;
         }
+
+        Object mergedValue = entry.getValue();
+        setMergedValue(newRecord, column, mergedValue);
+      }
+    }
+
+    // handle comparison columns
+    for (String column: _comparisonColumns) {
+      if (newRecord.isNullValue(column) && !prevRecord.isNullValue(column)) {
+        newRecord.putValue(column, prevRecord.getValue(column));
+        newRecord.removeNullValueField(column);
       }
     }
   }
+
+  private void setMergedValue(GenericRow newRecord, String column, Object mergedValue) {
+    if (mergedValue != null) {
+      // remove null value field if it was set
+      newRecord.removeNullValueField(column);
+      newRecord.putValue(column, mergedValue);
+    } else {
+      // if column exists but mapped to a null value then merger result was a null value
+      newRecord.addNullValueField(column);
+      newRecord.putValue(column, null);
```

Comment:
(MAJOR) This won't work. In order to set a value to null, you'll need to call `putDefaultNullValue()` with a default null value. Do you see a scenario where you want to explicitly set a value to `null`?

Reply:
I might not be fully aware of the null value support feature, let me update it with `putDefaultNullValue()`. Thank you. I don't forsee a scenario for null value.